### PR TITLE
Support setting column types in PostgreSQL connector

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Relates to #15515

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# PostgreSQL
* Add support for changing column types. ({issue}`15515`)
```

### CODE REVIEW

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 1]
```diff
@@ -2186,6 +2187,147 @@ public void testRenameColumn()
...
+        }
+    }
+
+    @Test
+    public void testSetColumnTypeWithComment()
+    {
+        skipTestUnless(hasBehavior(SUPPORTS_SET_COLUMN_TYPE) && hasBehavior(SUPPORTS_CREATE_TABLE_WITH_COLUMN_COMMENT));
+
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_set_column_type_comment_", "(col int COMMENT 'test')")) {
+            assertEquals(getColumnComment(table.getName(), "col"), "test");
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2275]: "my test comment" (or something like that)

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 2]
```diff
@@ -2186,6 +2187,147 @@ public void testRenameColumn()
...
+                .add(new SetColumnTypeSetup("integer", "bigint", "2147483647", "bigint '2147483647'"))
+                .add(new SetColumnTypeSetup("bigint", "integer", "-2147483648", "-2147483648"))
+                .add(new SetColumnTypeSetup("decimal(5,3)", "decimal(38,3)", "12.345", "12.345"))
+                .add(new SetColumnTypeSetup("date", "date", "date '2023-01-04'", "date '2023-01-04'"))
+                .add(new SetColumnTypeSetup("time(3)", "time(6)", "time '15:03:00.123'", "time '15:03:00.123000'"))
+                .add(new SetColumnTypeSetup("timestamp(3)", "timestamp(6)", "timestamp '2020-02-12 15:03:00.123'", "timestamp '2020-02-12 15:03:00.123000'"))
+                .add(new SetColumnTypeSetup("timestamp(3) with time zone", "timestamp(6) with time zone", "timestamp '2020-02-12 15:03:00.123 +01:00'", "timestamp '2020-02-12 15:03:00.123000 +01:00'"))
+                .add(new SetColumnTypeSetup("char(100)", "varchar", "'char-to-varchar'", "'char-to-varchar'"))
+                .add(new SetColumnTypeSetup("varchar", "char(100)", "'varchar-to-char'", "cast('varchar-to-char' as char(100))"))
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2249]: add char & varchar tests with length reduction

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 3]
```diff
@@ -2186,6 +2187,147 @@ public void testRenameColumn()
...
+                    .skippingTypesCheck()
+                    .matches("VALUES " + setup.newValueLiteral);
+        }
+        finally {
+            assertUpdate("DROP TABLE IF EXISTS " + tableName);
+        }
+    }
+
+    @DataProvider
+    public Object[][] setColumnTypesProvider()
```
Comment (findepi) [inline testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java:2223]: Provider -> DataProvider (at least that's my convention)

**Review by findepi — APPROVED** [review]

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorTest.java [hunk 4]
```diff
@@ -70,6 +70,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_DELETE:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
+                return false;
+
```
Comment (hashhar) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorTest.java:75]: above SUPPORTS_DELETE?

File: plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerConnectorTest.java [hunk 5]
```diff
@@ -87,6 +87,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_COMMENT_ON_COLUMN:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerConnectorTest.java:90]: together with `ADD_COLUMN_WITH_COMMENT`?

File: plugin/trino-singlestore/src/test/java/io/trino/plugin/singlestore/TestSingleStoreConnectorTest.java [hunk 6]
```diff
@@ -96,6 +96,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_COMMENT_ON_COLUMN:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-singlestore/src/test/java/io/trino/plugin/singlestore/TestSingleStoreConnectorTest.java:99]: together with ADD_COLUMN_WITH_COMMENT

File: plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java [hunk 7]
```diff
@@ -70,6 +70,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_CREATE_TABLE_WITH_COLUMN_COMMENT:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-redshift/src/test/java/io/trino/plugin/redshift/TestRedshiftConnectorTest.java:73]: together with ADD_COLUMN_WITH_COMMENT?

File: plugin/trino-raptor-legacy/src/test/java/io/trino/plugin/raptor/legacy/BaseRaptorConnectorTest.java [hunk 8]
```diff
@@ -86,6 +86,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_ADD_COLUMN_WITH_COMMENT:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-raptor-legacy/src/test/java/io/trino/plugin/raptor/legacy/BaseRaptorConnectorTest.java:89]: together with ADD_COLUMN_WITH_COMMENT

File: plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java [hunk 9]
```diff
@@ -105,6 +105,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_ADD_COLUMN_WITH_COMMENT:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:108]: together with ADD_COLUMN_WITH_COMMENT?

File: plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java [hunk 10]
```diff
@@ -76,6 +76,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_ADD_COLUMN_WITH_COMMENT:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-oracle/src/test/java/io/trino/plugin/oracle/BaseOracleConnectorTest.java:79]: together with ADD_COLUMN_WITH_COMMENT

File: plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/BaseMySqlConnectorTest.java [hunk 11]
```diff
@@ -78,6 +78,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_ADD_COLUMN_WITH_COMMENT:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-mysql/src/test/java/io/trino/plugin/mysql/BaseMySqlConnectorTest.java:81]: together with ADD_COLUMN_WITH_COMMENT

File: plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java [hunk 12]
```diff
@@ -66,6 +66,9 @@ protected boolean hasBehavior(TestingConnectorBehavior connectorBehavior)
             case SUPPORTS_ADD_COLUMN_WITH_COMMENT:
                 return false;
 
+            case SUPPORTS_SET_COLUMN_TYPE:
```
Comment (hashhar) [inline plugin/trino-mariadb/src/test/java/io/trino/plugin/mariadb/BaseMariaDbConnectorTest.java:69]: together with ADD_COLUMN_WITH_COMMENT

**Review by hashhar — COMMENTED** [review]
I don't have useful comments - just about enum ordering. Sorry.

Looks good otherwise.

File: plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java [hunk 13]
```diff
@@ -907,6 +908,15 @@ public void renameColumn(ConnectorSession session, ConnectorTableHandle table, C
         jdbcClient.renameColumn(session, tableHandle, columnHandle, target);
     }
 
+    @Override
+    public void setColumnType(ConnectorSession session, ConnectorTableHandle table, ColumnHandle column, Type type)
+    {
+        JdbcTableHandle tableHandle = (JdbcTableHandle) table;
+        JdbcColumnHandle columnHandle = (JdbcColumnHandle) column;
+        verify(!tableHandle.isSynthetic(), "Not a table reference: %s", tableHandle);
+        jdbcClient.setColumnType(session, tableHandle, columnHandle, type);
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:917]: The context of existing full handle seems that third-party developers may use `authorization` field. Let me keep as-is in this PR.

[the remark this answers is not in the document]
Comment (hashhar) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:917]: cc: @Praveen2112 It seems if authorization field is non-empty we should fail all operations?

[the remark this answers is not in the document]
Comment (Praveen2112) [reply plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/DefaultJdbcMetadata.java:917]: Yes.

File: testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java [hunk 14]
```diff
@@ -2186,6 +2187,147 @@ public void testRenameColumn()
...
+    {
+        return setColumnTypeSetupData().stream()
+                .map(this::applyColumnMapping)
+                .collect(toDataProvider());
+    }
+
+    protected SetColumnTypeSetup applyColumnMapping(SetColumnTypeSetup setup)
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply testing/trino-testing/src/main/java/io/trino/testing/BaseConnectorTest.java]: > how a connector can say this particular conversion isn't support in this connector case? (ie turn positive test case into a negative one)

Let me handle this in follow-up as the logic is unreachable in this PR. 

> how a connector can say this particular conversion is problematic, needs to be skipped (TODO + issue)?

Changed to optional type.

**Comment by colebow** [comment]
@ebyhr should we add any docs for this?

**Comment by ebyhr** [comment]
@colebow We need to refactor the docs first as the current granularity is `ALTER TABLE`.
https://trino.io/docs/current/connector/postgresql.html#alter-table

**Pull request merged**